06_gpu_and_ml/stable_lm/main.py

```diff
@@ -0,0 +1,146 @@
+import time
+from pathlib import Path
+
+import torch
+from huggingface_hub import snapshot_download
+from pydantic import BaseModel, Field
+from rich import print
+from transformers import (
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    GenerationConfig,
+    StoppingCriteria,
+    StoppingCriteriaList,
+    pipeline,
+)
+
+import modal
+
+
+def fetch_models():
+    MODEL_PATH = snapshot_download("stabilityai/stablelm-tuned-alpha-7b", ignore_patterns=["*.md"], revision="25071b093c15c0d1cb2b2876c6deb621b764fcf5")
+
+
+def compile_models():
+    from transformers import AutoModelForCausalLM, AutoTokenizer
+    MODEL_PATH = snapshot_download("stabilityai/stablelm-tuned-alpha-7b", ignore_patterns=["*.md"], revision="25071b093c15c0d1cb2b2876c6deb621b764fcf5")
+    m = AutoModelForCausalLM.from_pretrained(MODEL_PATH, torch_dtype=torch.float16, local_files_only=True).cuda()
+    tok = AutoTokenizer.from_pretrained(MODEL_PATH, local_files_only=True)
+    m = torch.compile(m)
+    m.save_pretrained(MODEL_PATH, safe_serialization=True, max_shard_size="24GB")
+    tok.save_pretrained(MODEL_PATH)
+    [p.unlink() for p in Path(MODEL_PATH).rglob("*.bin")]
+
+
+image = (
+    modal.Image.conda()
+    .apt_install("git", "software-properties-common", "curl", "wget", "libopenblas-dev", "libblas-dev", "g++", "libboost-all-dev", "cmake", "ninja-build", "libpq-dev", "libatlas-base-dev", "gfortran", "libclblast-dev", "libprotobuf-dev", "protobuf-compiler")
+    .conda_install(
+        "cudatoolkit-dev=11.7",
+        "pytorch-cuda=11.7",
+        channels=["nvidia", "pytorch"],
+    )
+    .env({"HF_HOME": "/root", "SAFETENSORS_FAST_GPU": "1", "BITSANDBYTES_NOWELCOME": "1", "PIP_DISABLE_PIP_VERSION_CHECK": "1", "PIP_NO_CACHE_DIR": "1"})
+    .run_commands(
+        "pip install transformers==4.28.1 accelerate==0.18.0 safetensors==0.3.0 bitsandbytes==0.38.1 sentencepiece==0.1.98",
+        gpu="A10G"
+    )
+    .run_function(
+        fetch_models,
+        gpu=None,
+        timeout=3600,
+    )
+    .run_function(
+        compile_models,
+        gpu="A10G",
+        timeout=3600,
+    )
```

Comment:
```suggestion
    .run_function(
        compile_models,
        gpu="A10G",
        timeout=3600,
    )
```

I think here you're trying to avoid using a GPU on the step which is just network-bound, but splitting into two functions makes the build less efficient. What's happening under the hood is that the first function runs in a new container the resulting filesystem state is baked as an image layer. Then the second function starts a new container with the previous image layer as a base. But in order to use the previously built image layer as a base it needs to download it. So this is still downloading the snapshot twice: once from Huggingface CDN and once from within Modal.

Reply:
All makes sense, thanks for the explanation! We can combine these.